File: Simulators/simulators/api/system/networkDeviceFunction.js
/*
 * Copyright (c) 2015-2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const BaseAsset = require('../common/baseAsset').BaseAsset;
const httpStatusCode = require('../../utils/httpStatusCode').httpStatusCode;

class NetworkDeviceFunction extends BaseAsset {
    constructor(mock, system, id) {
        super(mock, system, id);

        this.Ethernet = {
            MACAddress: 'E9:47:D3:60:64:66'
        };

        this.deviceEnabled = true;
        this.iSCSIBoot = {
            IPAddressType: 'IPv4',
            InitiatorIPAddress: '159.112.110.1',
            InitiatorName: 'Initiator Name',
            InitiatorDefaultGateway: 'Default gateway',
            InitiatorNetmask: '255.255.255.1',
            TargetInfoViaDHCP: false,
            PrimaryTargetName: 'Primary target name',
            PrimaryTargetIPAddress: '159.112.111.1',
            PrimaryTargetTCPPort: 66,
            PrimaryLUN: 10,
            PrimaryVLANEnable: true,
            PrimaryVLANId: 11,
            PrimaryDNS: '7.7.7.7',
            SecondaryTargetName: 'Secondary target name',
            SecondaryTargetIPAddress: '159.112.112.1',
            SecondaryTargetTCPPort: 77,
            SecondaryLUN: 12,
            SecondaryVLANEnable: false,
            SecondaryVLANId: 13,
            SecondaryDNS: '6.6.6.6',
            IPMaskDNSViaDHCP: false,
            RouterAdvertisementEnabled: true,
            AuthenticationMethod: 'CHAP',
            CHAPUsername: 'NDF CHAP username',
            CHAPSecret: 'nooneknows',
            MutualCHAPUsername: 'NDF Mutual CHAP username',
            MutualCHAPSecret: 'mutualnooneknows',
        }
        this.mock = mock;
        this.system = system;
        this.id = id;
        this.MockGetWithDynamicBody();
        this.MockPatch();
    }

    GetBaseUri() {
        return super.GetBaseUri('NetworkDeviceFunctions');
    }

    CreateJsonTemplate() {
        return {
            '@odata.context': '/redfish/v1/$metadata#Systems/Members/1/NetworkInterfaces/Members/1/NetworkDeviceFunctions/Members/$entity',
            '@odata.id': this.GetBaseUri(),
            '@odata.type': '#NetworkDeviceFunction.v1_0_0.NetworkDeviceFunction',
            'Name': 'NetworkDeviceFunction',
            'Id': this.Id,
            'Description' : 'description',
            'Status': {
                'State': 'Enabled',
                'Health': 'OK'
            },
            'DeviceEnabled': this.deviceEnabled,
            'Ethernet': this.Ethernet,
            'iSCSIBoot': this.iSCSIBoot,
            'Links': {}
        };
    }

    MockPatch() {
        this.mock.mockPatch(this.GetBaseUri(),
            httpStatusCode.OK,
            undefined,
            (req, response) => {
                var jsonRequest = JSON.parse(req.body);

                this.deviceEnabled = jsonRequest.DeviceEnabled === undefined ? this.deviceEnabled : jsonRequest.DeviceEnabled;
                if (jsonRequest.iSCSIBoot !== null) {
                    for (let property in jsonRequest.iSCSIBoot) {
                        this.iSCSIBoot[property] = jsonRequest.iSCSIBoot[property];
                    }
                }
                else {
                    for (let property in this.iSCSIBoot) {
                        this.iSCSIBoot[property] = null;
                    }
                }

                if (jsonRequest.Ethernet !== null) {
                    for (let property in jsonRequest.Ethernet) {
                        this.Ethernet[property] = jsonRequest.Ethernet[property];
                    }
                }
                else {
                    this.Ethernet.MACAddress = null;
                }

                response.body = this.CreateJsonTemplate();
                return response;
            }
        );
    }
}

exports.NetworkDeviceFunction = NetworkDeviceFunction;

File: PSME/application/src/rest/validators/schemas/network_device_function.cpp
/*!
 * @brief Definition of Network Device Function PATCH request validator
 *
 * @copyright Copyright (c) 2017-2019 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file network_device_function.cpp
 */

#include "psme/rest/validators/schemas/network_device_function.hpp"
#include "psme/rest/validators/schemas/common.hpp"
#include "psme/rest/constants/constants.hpp"
#include "agent-framework/module/enum/common.hpp"
#include "agent-framework/module/enum/compute.hpp"


using namespace psme::rest;
using namespace psme::rest::validators::schema;
using namespace agent_framework::model;



const jsonrpc::ProcedureValidator& NetworkDeviceFunctionPatchSchema::EthernetObjectSchema::get_procedure() {
    static jsonrpc::ProcedureValidator procedure{
        jsonrpc::PARAMS_BY_NAME,
        constants::Common::MAC_ADDRESS, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        nullptr
    };
    return procedure;
}


const jsonrpc::ProcedureValidator& NetworkDeviceFunctionPatchSchema::IscsiBootObjectSchema::get_procedure() {
    static jsonrpc::ProcedureValidator procedure{
        jsonrpc::PARAMS_BY_NAME,
        constants::NetworkDeviceFunction::IP_ADDRESS_TYPE, VALID_OPTIONAL(VALID_NULLABLE(VALID_ENUM(enums::IPAddressType))),
        constants::NetworkDeviceFunction::INITIATOR_IP_ADDRESS, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::INITIATOR_NAME, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::INITIATOR_DEFAULT_GATEWAY, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::INITIATOR_NETMASK, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::TARGET_INFO_VIA_DHCP, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_BOOLEAN)),
        constants::NetworkDeviceFunction::PRIMARY_TARGET_NAME, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::PRIMARY_TARGET_IP_ADDRESS, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::PRIMARY_TARGET_TCP_PORT, VALID_OPTIONAL(VALID_NULLABLE(VALID_NUMERIC_TYPED(UINT32))),
        constants::NetworkDeviceFunction::PRIMARY_LUN, VALID_OPTIONAL(VALID_NULLABLE(VALID_NUMERIC_EQGT(INT64,0))),
        constants::NetworkDeviceFunction::PRIMARY_VLAN_ENABLE, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_BOOLEAN)),
        constants::NetworkDeviceFunction::PRIMARY_VLAN_ID, VALID_OPTIONAL(VALID_NULLABLE(VALID_NUMERIC_TYPED(UINT32))),
        constants::NetworkDeviceFunction::PRIMARY_DNS, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::SECONDARY_TARGET_NAME, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::SECONDARY_TARGET_IP_ADDRESS, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::SECONDARY_TARGET_TCP_PORT, VALID_OPTIONAL(VALID_NULLABLE(VALID_NUMERIC_TYPED(UINT32))),
        constants::NetworkDeviceFunction::SECONDARY_LUN, VALID_OPTIONAL(VALID_NULLABLE(VALID_NUMERIC_EQGT(INT64,0))),
        constants::NetworkDeviceFunction::SECONDARY_VLAN_ENABLE, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_BOOLEAN)),
        constants::NetworkDeviceFunction::SECONDARY_VLAN_ID, VALID_OPTIONAL(VALID_NULLABLE(VALID_NUMERIC_TYPED(UINT32))),
        constants::NetworkDeviceFunction::SECONDARY_DNS, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::IP_MASK_DNS_VIA_DHCP, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_BOOLEAN)),
        constants::NetworkDeviceFunction::ROUTER_ADVERTISEMENT_ENABLED, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_BOOLEAN)),
        constants::NetworkDeviceFunction::AUTHENTICATION_METHOD, VALID_OPTIONAL(VALID_NULLABLE(VALID_ENUM(enums::FunctionAuthenticationMethod))),
        constants::NetworkDeviceFunction::CHAP_USERNAME, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::CHAP_SECRET, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::MUTUAL_CHAP_USERNAME, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        constants::NetworkDeviceFunction::MUTUAL_CHAP_SECRET, VALID_OPTIONAL(VALID_NULLABLE(VALID_JSON_STRING)),
        nullptr
    };
    return procedure;
}


const jsonrpc::ProcedureValidator& NetworkDeviceFunctionPatchSchema::get_procedure() {
    static jsonrpc::ProcedureValidator procedure{
        jsonrpc::PARAMS_BY_NAME,
        constants::NetworkDeviceFunction::ETHERNET, VALID_OPTIONAL(VALID_ATTRIBUTE(EthernetObjectSchema)),
        constants::NetworkDeviceFunction::ISCSI_BOOT, VALID_OPTIONAL(VALID_ATTRIBUTE(IscsiBootObjectSchema)),
        // TODO:validate IP addresses and MAC addresses
        nullptr
    };
    return procedure;
}
